Custom list for cards

qnkhuat opened the pull request:
Implement Milestone 2 for #26873 - supports custom list for Card.

Like Custom list on dashboard, parameters on cards now will have 2 new keys: `values_source_type` and `values_source_config` 

for source=static-list
```diff
{:name                 "Static Category",
 :slug                 "static_category"
 :id                   "_STATIC_CATEGORY_",
 :type                 "category",
+ :values_source_type   "static-list"
+ :values_source_config {:values ["African" "American" "Asian"]}
```
For source=card
```diff
{:name                 "Static Category",
 :slug                 "static_category"
 :id                   "_STATIC_CATEGORY_",
 :type                 "category",
+ :values_source_type   "card"
+ :values_source_config {:card_id     source-card-id
+                        :value_field [:field 3 nil]}}
```

### API
Since we haven't had any API to fetch values for parameter on card, this PR add 2 new endpoints similar to what we have on dashboard:
- `/api/card/:card-id/params/:param-key/values` - to get the values list
- `/api/card/:card-id/params/:param-key/search/:query` - to search for values
- `/api/public/card/:card-uuid/params/:param-key/values` - to get the values list on public card
- `/api/public/card/:card-uuid/params/:param-key/search/:query` - to search for values on public card
Note: I intentionally skip writing tests for permissions, those tests will be added in https://github.com/metabase/metabase/pull/27439 where I add permissions tests for both custom list on dashboard and card

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/27403)
<!-- Reviewable:end -->

qnkhuat committed:
custom list for cards and test

qnkhuat committed:
whitespace lint and docs

qnkhuat committed:
does this break serdes?

qnkhuat committed:
more extract tests

qnkhuat committed:
a lot more serdes tests

qnkhuat committed:
revert removing ingest-path for card

qnkhuat committed:
wording and add tests for dashboard descendent when it has parameter_mappings on dashcard

qnkhuat committed:
fix path serdes for card tests

qnkhuat committed:
whitespace

qnkhuat committed:
CRUD on card.parameters will also update ParameterCard

qnkhuat committed:
clj-kondo

qnkhuat committed:
BE

tsmacdonald commented on src/metabase/models/parameter_card.clj:
```diff
@@ -45,27 +47,30 @@
+(s/defn upsert-or-delete-from-parameters!
+  "From a parameters list on card or dashboard, create, update,
+  or delete appropriate ParameterCards for each parameter in the dashboard"
+  [parameterized-object-type :- su/NonBlankString
```
should we use the `valid-parameterized_object_type` set to lock down this schema properly?

tsmacdonald commented on src/metabase/models/parameter_card.clj:
```diff
@@ -2,13 +2,15 @@
 
 (defonce ^{:doc "Set of valid parameterized_object_type for a ParameterCard"}
-  valid-parameterized_object_type #{"dashboard"})
+  valid-parameterized_object_type #{"dashboard" "card"})
```
I understand why it happened, but I'm a little :grimacing: about `-`/`_` mixing and probably would've used hyphens throughout.

qnkhuat replied:
haha yea I copied the whole word from under. Will change.

tsmacdonald commented on src/metabase/models/params/card_values.clj:
```diff
@@ -10,9 +10,9 @@
   "Maximum number of rows returned when running a card.
-  It's 2000 because this is the limit we use when viewing a question.
+  It's 1000 because it matches with the limit for chain-filter.
   Maybe we should lower it for the sake of display a parameter dropdown."
```
```suggestion
  Maybe we should lower it for the sake of displaying a parameter dropdown."
```

qnkhuat replied:
thank!

tsmacdonald commented on src/metabase/api/card.clj:
```diff
@@ -930,4 +933,48 @@ saved later when it is ready."
+       "card"        (params.card-values/param->values param query)
+       (throw (ex-info (tru "Invalid values-source-type: {0}" (pr-str source-type))
+                       {:values-source-type source-type
+                        :status-code        400}))))))
```
I wonder if the whole `case` should be extracted into a common ns. If we added another source-type it'd be too bad to have to update the dispatcher in both the card and dashboard APIs. Unfortunately with all the card-specific error messaging we can't get too DRY, but I think extracting the main branching is a good result.

qnkhuat replied:
I thought about it too, but the card does not support chain-filter and it makes things awkward.

Another idea is use defmethod that dispatch with `parameterized-object-type, values-source-type`.

but I think it's a bit too much for now. Definitely need to revisit this idea once we support custom-list on data apps.

tsmacdonald reviewed (approved):
left some suggestions but nothing mission-critical and overall it looks great to me, thanks Ngoc!

qnkhuat committed:
resolve Tim's suggestions

ranquild reviewed (commented):
Seems good but hard to tell for sure until we use it on the FE. Do you want to merge it now or wait for that?

qnkhuat commented:
I'm having 3 PRs based on this branch so I'd like to merge it asap.
I think we can always make changes if FE finds anything weird about it?

dpsutton commented on src/metabase/models/parameter_card.clj:
```diff
@@ -3,19 +3,21 @@
 
 (models/defmodel ParameterCard :parameter_card)
 
 (defonce ^{:doc "Set of valid parameterized_object_type for a ParameterCard"}
```
just randomly but why is this a `defonce`?

qnkhuat replied:
just thought it gives a bit of performance to define constants with defonce. but I guess it's not dev friendly

dpsutton replied:
no performance aspect. just doesn't reload

```clojure
(defmacro defonce
  "defs name to have the root value of the expr iff the named var has no root value,
  else expr is unevaluated"
  {:added "1.0"}
  [name expr]
  `(let [v# (def ~name)]
     (when-not (.hasRoot v#)
       (def ~name ~expr))))
```

So the only thing that it ultimately does is `(def thing val)` with a conditional to not do that if it already exists.

ranquild reviewed (changes requested):
@qnkhuat We're currently calling `/api/field/{id}/values` for template tags. Should the frontend use new endpoints in all cases?

dpsutton commented on src/metabase/api/card.clj:
```diff
@@ -930,4 +932,48 @@ saved later when it is ready."
+  (let [card (api/read-check Card card-id)]
+    (param-values card param-key)))
+
+(api/defendpoint GET "/:card-id/params/:param-key/search/:query"
```
Why two api endpoints here? I would think that we could do something like
`GET api/<card-id>/params/<param-key>/values?search=foo` to handle both? As far as I can tell from the bodies of these, they are the same except for the inclusion or not of `query`. Would it make sense to collapse these into one and just check for a `?search=<search-query>` query param?

qnkhuat replied:
this is to be asymmetry with ho`api/dashboard`. we could use one single API if you think it makes more sense.

ranquild replied:
Would be happy to have a single endpoint instead of a pair. We could also consolidate  `api/dashboard/:dashId/params/:paramId/values` and `search` endpoints into just one endpoint.

qnkhuat committed:
add public endpoints to fetch values for card and tests

qnkhuat committed:
fix conflicts

qnkhuat committed:
fix clj-kondo

qnkhuat committed:
fix the annoying clj-kondo unresolved-symbol on macro

qnkhuat committed:
fix failed to run tests in ee

qnkhuat committed:
resolve Braden's comments

qnkhuat commented:
I accidentally merged a [PR](https://github.com/metabase/metabase/pull/27422) that adds serdes for custom-list into this PR. (Braden approved it tho)

So if you see anything serdes related in this PR, maybe just ignore it or use this [PR](https://github.com/metabase/metabase/pull/27422) to review about serdes.

qnkhuat committed:
fix schema

qnkhuat committed:
fix serdes tests and paramter schema

ranquild reviewed (changes requested):
1. When I call these endpoints I get an error `Invalid values-source-type: nil`. I expect this should work for all source types, including the default one. This would also match how it works in dashboards.
2. I also need `embed` versions of endpoints.

dpsutton committed:
more tests

dpsutton committed:
fixups from rebase

dpsutton committed:
Hook up card params endpoint for current style as well as new

Previously the new `"/:card-id/params/:param-key/values"` endpoint
handled the "static-list" and the "card" new styles of parameters. But
it did not handle the "current" style to look up field values. The FE
was expected to hit `api/field/:id/values`.

This change makes the card endpoint work more like the dashboard
endpoint which did handle all three. It cannot be exactly the same
though as we aren't looking on dashboard's for params and we don't have
dashcards.

So we have to skip a lot of the chain filter stuff but we end up using
the same stuff at the end.

dpsutton committed:
lint fixes

dpsutton committed:
expose embedded param values for cards

calherries committed:
Merge branch 'master' into ngoc-custom-values-with-card

ranquild committed:
Fix duplicate import
